File: app/api/src/Astra.Api/Llm/Archetypes/java-spring/canonical-php-catalog-service/src/main/java/com/acme/catalog/MoneyMath.java
// SPDX-Spec: php/invoice.php (signed)
// SPDX-Archetype: canonical-php-catalog-service
package com.acme.catalog;

import java.math.BigDecimal;
import java.math.RoundingMode;

/**
 * Java 21 projection of PHP {@code invoice.php}'s {@code lineTotal}.
 * The PHP source computed money in a float and compared it with a loose
 * {@code $total == "0"}; both are traps this target closes:
 *
 * <ul>
 *   <li>EC-1: money is a {@link BigDecimal} (never a double) so 0.1 + 0.2 == 0.3
 *       holds; the result is rounded HALF_UP to 2 decimal places.</li>
 *   <li>LTC-1: the loose {@code == "0"} becomes an explicit
 *       {@link BigDecimal#compareTo} against zero, not a string juggle.</li>
 * </ul>
 */
public final class MoneyMath {

    private MoneyMath() {}

    /** {@code $total = round($price * $qty, 2)} — as exact decimal money. */
    @SpecClaim("EC-1")
    @SpecClaim("INV-1")
    @SpecClaim("LTC-1")
    @TargetMapping(value = "BigDecimal arithmetic, HALF_UP scale 2",
                   phpConstruct = "round($price * $qty, 2) with float money")
    public static BigDecimal lineTotal(BigDecimal price, int qty) {
        if (price == null) {
            throw new IllegalArgumentException("price is required");
        }
        if (qty < 0) {
            throw new IllegalArgumentException("qty must be non-negative");
        }
        BigDecimal total = price.multiply(BigDecimal.valueOf(qty));
        // LTC-1: the PHP `if ($total == "0")` becomes a typed compareTo, not a
        // string juggle. (compareTo, not equals — 0 vs 0.00 must be equal.)
        if (total.compareTo(BigDecimal.ZERO) == 0) {
            return BigDecimal.ZERO.setScale(2, RoundingMode.HALF_UP);
        }
        return total.setScale(2, RoundingMode.HALF_UP);
    }
}

File: app/api/src/Astra.Api/Llm/Archetypes/dotnet8/canonical-php-catalog-service/src/MoneyMath.cs
// SPDX-Spec: php/invoice.php (signed)
// SPDX-Archetype: canonical-php-catalog-service (dotnet8)
namespace Acme.Catalog;

/// <summary>
/// C# projection of PHP invoice.php's lineTotal. Closes two PHP traps:
/// EC-1 money is a <see cref="decimal"/> (never a double) rounded to 2 dp
/// AwayFromZero; LTC-1 the loose <c>$total == "0"</c> becomes an explicit typed
/// comparison, not a string juggle.
/// </summary>
public static class MoneyMath
{
    /// <summary><c>round($price * $qty, 2)</c> as exact decimal money.</summary>
    [SpecClaim("EC-1")]
    [SpecClaim("INV-1")]
    [SpecClaim("LTC-1")]
    [TargetMapping("decimal arithmetic, MidpointRounding.AwayFromZero, 2 dp",
        PhpConstruct = "round($price * $qty, 2) with float money")]
    public static decimal LineTotal(decimal price, int qty)
    {
        if (qty < 0)
            throw new ArgumentException("qty must be non-negative", nameof(qty));

        decimal total = price * qty;
        // LTC-1: the PHP `if ($total == "0")` becomes a typed compare, not a juggle.
        if (total == 0m)
            return 0m;
        return Math.Round(total, 2, MidpointRounding.AwayFromZero);
    }
}
